bin/test_with_coverage.dart

```diff
@@ -0,0 +1,177 @@
+// Copyright (c) 2022, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:io';
+import 'dart:convert' show utf8, LineSplitter;
+
+import 'package:args/args.dart';
+import 'package:package_config/package_config.dart';
+import 'package:path/path.dart' as path;
+import 'package:coverage/src/util.dart' show extractVMServiceUri;
+
+final allProcesses = <Process>[];
+
+Future<void> dartRun(List<String> args,
+    {Function(String)? onStdout, String? workingDir}) async {
+  final process = await Process.start(
+    Platform.executable,
+    args,
+    workingDirectory: workingDir,
+  );
+  allProcesses.add(process);
+  final broadStdout = process.stdout.asBroadcastStream();
+  broadStdout.listen(stdout.add);
+  broadStdout
+      .transform(utf8.decoder)
+      .transform(const LineSplitter())
+      .listen(onStdout);
+  process.stderr.listen(stderr.add);
+  final result = await process.exitCode;
+  if (result != 0) {
+    throw ProcessException(Platform.executable, args, '', result);
+  }
+}
+
+Future<String?> getPackageName(String packageDir) async {
+  final config = await findPackageConfig(Directory(packageDir));
+  return config?.packageOf(Uri.directory(packageDir))?.name;
+}
+
+Future<void> main(List<String> arguments) async {
```

Comment:
This method is 135 lines long, consider extracting some sub methods, or moving some of the local methods to top level methods.